partiql-lib/partiql-sprout/src/main/kotlin/org/partiql/sprout/parser/ion/IonTypeParser.kt

```diff
@@ -171,57 +194,78 @@ internal object IonTypeParser : SproutParser {
+                        // Import type reference using '.'
+                        TypeRef.Import(symbol, nullable)
+                    }
+                    else -> {
+                        error("type reference `$symbol` not found")
+                    }
+                }
+            }
+            // 2. Attempt as scalar
             try {
                 return TypeRef.Scalar(
                     type = ScalarType.valueOf(symbol.toUpperCase()),
                     nullable = nullable,
                 )
-            } catch (_: IllegalArgumentException) {}
-            // 2. Attempt to find the symbol in the definitions
-            val node = if (absolute) {
-                val path = symbol.split(".")
-                root.search(path)
-            } else {
-                tip.search(symbol)
+            } catch (_: IllegalArgumentException) {
             }
+            // 3. Attempt to find the symbol relative to the current position
+            val node = tip.search(symbol)
             if (node != null) {
                 return TypeRef.Path(
                     nullable = nullable,
                     ids = (node.path.toTypedArray()),
                 )
             }
-            // 3. Attempt to find the symbol in the imports
+            // 4. Attempt to find the symbol in the imports
             if (imports.symbols.contains(symbol)) {
                 return TypeRef.Import(symbol, nullable)
             }
-            // 4. Error nothing found
-            error("symbol `$symbol` not found")
+            // 5. Error nothing found
+            error("type reference `$symbol` not found")
         }
 
         /**
          * Resolve the collection type
          */
         private fun resolve(v: IonList): TypeRef {
-            val (symbol, _, nullable) = symbol(v.id())
+            val (symbol, nullable) = v.ref()
             return when (symbol.toLowerCase()) {
                 "list" -> {
-                    assert(v.size == 1)
+                    assert(v.size == 1) { "list must have exactly one type" }
+                    // assert(v[0] is IonSymbol) { "list type parameter must be a symbol" }
                     val t = resolve(v[0])
                     TypeRef.List(t, nullable)
                 }
                 "set" -> {
-                    assert(v.size == 1)
+                    assert(v.size == 1) { "set must have exactly one type" }
+                    // assert(v[0] is IonSymbol) { "set type parameter must be a symbol" }
                     val t = resolve(v[0])
                     TypeRef.Set(t, nullable)
                 }
                 "map" -> {
-                    assert(v.size == 2)
+                    assert(v.size == 2) { "map must have exactly two types" }
+                    // assert(v[0] is IonSymbol) { "map key type parameter must be a symbol" }
+                    // assert(v[1] is IonSymbol) { "map value type parameter must be a symbol" }
```

Comment:
It wasn't like this, but then I simplified for code generation. Now that both parsing and handles nested collections, then I commented these (should've deleted, but I wanted the visual reminder).

I initially had map keys as only scalars, but that's a bit limiting if the target has an appropriate map implementations available to use a node as a key. So now I've only limited the key to a symbol so that we aren't allowing inline definitions or collections.